config.go

```diff
@@ -27,19 +27,88 @@ type config struct {
 	DisableCheckpointSignature bool `json:"disable_checkpoint_signature"`
 	PluginMinPort              int
 	PluginMaxPort              int
-
-	Builders       packer.MapOfBuilder
-	Provisioners   packer.MapOfProvisioner
-	PostProcessors packer.MapOfPostProcessor `json:"post-processors"`
+	RawBuilders                map[string]string         `json:"builders"`
+	RawProvisioners            map[string]string         `json:"provisioners"`
+	RawPostProcessors          map[string]string         `json:"post-processors"`
+	Builders                   packer.MapOfBuilder       `json:"-"`
+	Provisioners               packer.MapOfProvisioner   `json:"-"`
+	PostProcessors             packer.MapOfPostProcessor `json:"-"`
 }
 
-// Decodes configuration in JSON format from the given io.Reader into
+// decodeConfig decodes configuration in JSON format from the given io.Reader into
 // the config object pointed to.
 func decodeConfig(r io.Reader, c *config) error {
 	decoder := json.NewDecoder(r)
 	return decoder.Decode(c)
 }
 
+// LoadExternalComponentsFromConfig loads plugins defined in RawBuilders, RawProvisioners, and RawPostProcessors.
+func (c *config) LoadExternalComponentsFromConfig() {
+	// helper to build up list of plugin paths
+	extractPaths := func(m map[string]string) []string {
+		paths := make([]string, 0, len(m))
+		for _, v := range m {
+			paths = append(paths, v)
+		}
+
+		return paths
+	}
+
+	var pluginPaths []string
+	pluginPaths = append(pluginPaths, extractPaths(c.RawProvisioners)...)
+	pluginPaths = append(pluginPaths, extractPaths(c.RawBuilders)...)
+	pluginPaths = append(pluginPaths, extractPaths(c.RawPostProcessors)...)
+
+	var externallyUsed = make([]string, 0, len(pluginPaths))
+	for _, pluginPath := range pluginPaths {
+		if name, ok := c.loadExternalComponent(pluginPath); ok {
+			log.Printf("[DEBUG] Loaded plugin: %s = %s", name, pluginPath)
+			externallyUsed = append(externallyUsed, name)
+		}
+	}
+
+	if len(externallyUsed) > 0 {
+		sort.Strings(externallyUsed)
+		log.Printf("using external builders %v", externallyUsed)
```

Comment:
Logline nit: is this actually for only builders or for all plugins?

Reply:
Ah, good catch. It should read `using external plugins`. Hmmm :thinking: do you think that will add confusion? 

It is nice that the discoverExternalComponents function groups each custom plugin by type.

Reply:
I don't think it's a blocker but if you want to try to customize you can.

Reply:
Totally agree. I gave it a run after making the log updates to make sure things look okay as they are. Thanks for reviewing and approving